src/Service/FirebaseMessageService.php

```diff
@@ -0,0 +1,222 @@
+        'normal',
+      ])) {
+      $this->body['priority'] = $options['priority'];
+    }
+    if (isset($options['content_available'])) {
+      $this->body['content_available'] = (bool) $options['content_available'];
+    }
+    if (isset($options['mutable_content'])) {
+      $this->body['mutable_content'] = (bool) $options['mutable_content'];
+    }
+    if (isset($options['time_to_live']) && ((int) $options['time_to_live'] >= 0 && (int) $options['time_to_live'] <= 2419200)) {
+      $this->body['time_to_live'] = $options['time_to_live'];
+    }
+    if (isset($options['dry_run'])) {
+      $this->body['dry_run'] = (bool) $options['dry_run'];
+    }
+  }
+
+  /**
+   * Add single device or group of devices or multiple devices to message target.
+   *
+   * @param string|array $recipients
+   *   Recipients of message.
+   */
+  public function setRecipients($recipients) {
+    if (is_array($recipients)) {
+      if (count($recipients) < self::MAX_DEVICES) {
+        $this->body['registration_ids'] = $recipients;
+      }
+      else {
+        throw new \OutOfRangeException(sprintf('Message device limit exceeded. Firebase supports a maximum of %u devices.', self::MAX_DEVICES));
+      }
+    }
+    else {
+      $this->body['to'] = $recipients;
+    }
+  }
+
+  /**
+   * Add topic or list of topics message target.
+   *
+   * @see https://firebase.google.com/docs/cloud-messaging/send-message#send_messages_to_topics
+   *
+   * @param string|array $topics
+   *   Topics without "/topics/".
+   */
+  public function setTopics($topics) {
+    if (is_array($topics) and count($topics) > 1) {
+      if (count($topics) > self::MAX_TOPICS) {
+        throw new \OutOfRangeException(sprintf('Topics limit exceeded. Firebase supports a maximum of %u topics.', self::MAX_TOPICS));
+      }
+      elseif (!$this->condition) {
+        throw new \InvalidArgumentException('Missing message condition. You must specify a condition pattern when sending to combinations of topics.');
+      }
+      elseif (count($topics) != substr_count($this->condition, '%s')) {
+        throw new \UnexpectedValueException('The number of message topics must match the number of occurrences of "%s" in the condition pattern.');
+      }
+      else {
+        foreach ($topics as &$topic) {
+          $topic = vsprintf("'%s' in topics", $topic);
+        }
+        $this->body['condition'] = vsprintf($this->condition, $topics);
+      }
+    }
+    else {
+      if (is_array($topics)) {
+        $topics = reset($topics);
+      }
+      $this->body['to'] = sprintf('/topics/%s', $topics);
+    }
+  }
+
+  /**
+   * Add notification to message.
+   *
+   * @param array $notification
+   *   Notification options.
+   */
+  public function setNotification(array $notification) {
+    if (!empty($notification['title']) && !empty($notification['body'])) {
+      $this->body['notification'] = $notification;
+    }
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function send() {
```

Comment:
will be cool to have some methods for delete "old" tokens

Reply:
we should have this (delete old tokens)
for send to group(s), to topic(s), to device tokens

for example https://firebase.google.com/docs/cloud-messaging/js/device-group
`Here is an example of "partial success" — the notification_key has 3 registration tokens associated with it. The message was successfully sent to 1 of the registration tokens only. The response message lists the registration tokens that failed to receive the message:

{
  "success":1,
  "failure":2,
  "failed_registration_ids":[
     "regId1",
     "regId2"
  ]
}`